src/hooks/useGlobalIntersectionObserverWithDelay.ts

```diff
@@ -0,0 +1,125 @@
+/* eslint-disable react-func/max-lines-per-function */
+import { useCallback, useRef } from 'react';
+
+import useBrowserLayoutEffect from '@/hooks/useBrowserLayoutEffect';
+
+const isDev = process.env.NEXT_PUBLIC_VERCEL_ENV === 'development';
+
+/**
+ * Sets a debugging text to element debugger node if we are in a dev environment
+ * and the element debugger paragraph node exists.
+ *
+ * @param {string} text
+ * @param {HTMLElement} intersectionDebuggerNode
+ *
+ */
+const setIntersectionDebuggerNodeText = (text: string, intersectionDebuggerNode: HTMLElement) => {
+  if (!isDev || !intersectionDebuggerNode) {
+    return;
+  }
+  // eslint-disable-next-line no-param-reassign
+  intersectionDebuggerNode.textContent = `OBSERVER DEBUGGER: ${text}`;
+};
+
+/**
+ * A hook that will create a global observer and inject it into window so that it's
+ * globally available. Once it's created, it can be used by any element that would like
+ * to be observed using that observer using {@see useObserverElement} hook. This hook
+ * is meant to be used for delayed element visibility callback.
+ *
+ * @param {IntersectionObserverInit} options
+ * @param {(element:Element)=>void} onElementVisible
+ * @param {string} observerId the name of the observer
+ * @param {string} elementIdKey the key of the unique identifier of each element.
+ * @param {string} delayForMS the number of milliseconds after which the observer should trigger the callback
+ */
+const useGlobalIntersectionObserverWithDelay = (
+  { threshold = 1, root = null, rootMargin = '0%' }: IntersectionObserverInit,
+  onElementVisible: (element: Element) => void,
+  observerId: string,
+  elementIdKey: string,
+  delayForMS: number,
+) => {
+  /**
+   * An object that will hold a reference of each unique identifier and
+   * the ID of the timeout function that will trigger when the element
+   * is visible for delayForMS number of seconds.
+   */
+  const timeouts = useRef<Record<any, number>>({});
```

Comment:
Let's use a `Map` instead, it's recommended to use maps over objects when you frequently add/delete elements ([reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps)).

Reply:
Sure will us Map instead.